src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -2,80 +2,120 @@
+                }
+            }
+        }
+        return instance;
+    }
+
+    @Override
+    public Optional<Product> getProduct(final Long id) {
+        Lock readLock = rwLock.readLock();
+        readLock.lock();
+        try {
+            return products.stream()
+                    .filter(product -> product.getId().equals(id))
+                    .findAny();
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public List<Product> findProducts(@NonNull final String searchQuery,
+                                      @NonNull final SortField sortField,
+                                      @NonNull final SortOrder sortOrder) {
+        Lock readLock = rwLock.readLock();
+        readLock.lock();
+        try {
+            Comparator<Product> comparator = Comparator.comparing(product -> {
+                if (sortField != null && SortField.description == sortField) {
+                    return (Comparable) product.getDescription();
+                } else {
+                    return (Comparable) product.getPrice();
+                }
+            });
+
+            if (sortOrder != null && SortOrder.desc == sortOrder) {
+                comparator = comparator.reversed();
+            }
+
+            return List.copyOf(products.stream()
+                    .filter((product -> searchQuery == null || searchQuery.isEmpty() ||
+                            product.getDescription().toUpperCase().contains(searchQuery.toUpperCase())))
+                    .filter(this::nonNullPrice)
+                    .filter(this::productIsInStock)
+                    .sorted(comparator)
+                    .collect(Collectors.toList()));
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    private boolean productIsInStock(@NonNull final Product product) {
+        return product.getStock() > 0;
+    }
+
+    private boolean nonNullPrice(@NonNull final Product product) {
+        return product.getPrice() != null;
+    }
+
+    @Override
+    public void save(@NonNull final Product product) {
+        Lock writeLock = rwLock.writeLock();
+        writeLock.lock();
+        try {
+            if (product.getId() != null) {
+                Optional<Product> optProduct = getProduct(product.getId());
+                optProduct.ifPresent(products::remove);
+            } else {
+                product.setId(maxId++);
+            }
+            products.add(product);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void delete(final Long id) {
+        Lock writeLock = rwLock.writeLock();
+        writeLock.lock();
+        try {
+            Optional<Product> optProduct = products.stream()
+                    .filter(product -> product.getId().equals(id))
+                    .findAny();
+            optProduct.ifPresent(products::remove);
+        } finally {
+            writeLock.unlock();
+        }
+    }
```

Comment:
please don't change format of the entire document
it's hard to see actual changes